Field stats: Added `format` option for index constraints

martijnvg opened the pull request:
Added a `format` for index constraints on date fields to allow dates of a different format to be specified than is defined in the date field's format setting in the mapping.

PR for #14804

rjernst commented on core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java:
```diff
@@ -449,7 +450,11 @@ public String getMaxValue() {
-            return dateFormatter.parser().parseMillis(value);
+            try {
+                return dateFormatter.parser().parseMillis(value);
+            } catch (Exception e) {
```
Can we support milliseconds through an explicit format parameter? We just got rid of this fallback mechanism in date fields for 2.0. It can be trappy when a user has no way to disable this builtin format.

martijnvg replied:
Ok, I'll remove the fallback logic and add a `format` option to the api that overrides the format that has been specified in the mapping for a date field.

bleskes replied:
+1 to *not* build in another hard coded fall back (we don't know what a number means - might be seconds). I do wonder what goes wrong here. The default date field formatter should fall back to mills:  `strict_date_optional_time||epoch_millis` .

martijnvg replied:
@bleskes This is in the case if someone has configured a custom date field format. (see #14804) If that happens Kibana doesn't know about this and it is very hard for them to convert to mills since epoch. So adding a format option would help here a lot.

martijnvg commented:
@rjernst @bleskes I've updated the PR and removed the hardcoded fallback logic with an optional `format` option that can be specified in the field stats api.

bleskes commented on core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java:
```diff
@@ -345,7 +346,7 @@ public void append(FieldStats stats) {
 
         @Override
-        protected java.lang.Double valueOf(String value) {
+        protected java.lang.Double valueOf(String value, String optionalFormat) {
```
this feels weird. We have a parameter and not use it. Can we add java docs and throw and operation not supported if someone used it? (this goes for all of those)

martijnvg replied:
I'll add the unsupported operation exception.

there is jdoc on the abstract method. I'll add there it is up to the implementation if the format parameter is used.

bleskes commented on core/src/main/java/org/elasticsearch/action/fieldstats/FieldStatsRequest.java:
```diff
@@ -121,22 +122,21 @@ private void parseIndexContraints(List<IndexConstraint> indexConstraints, XConte
+                                    comparison = IndexConstraint.Comparison.parse(parser.currentName());
+                                    value = parser.text();
+                                }
+                            }
```
can we add an else with an exception?

bleskes commented on core/src/main/java/org/elasticsearch/action/fieldstats/IndexConstraint.java:
```diff
@@ -61,6 +74,10 @@ public String getValue() {
         return value;
     }
 
+    public String getOptionalFormat() {
```
java docs?

bleskes commented on core/src/test/java/org/elasticsearch/fieldstats/FieldStatsTests.java:
```diff
@@ -359,4 +361,88 @@ public void testDateFiltering() {
         assertThat(response.getIndicesMergedFieldStats().get("test2").get("value").getMinValue(), equalTo("2014-01-02T00:00:00.000Z"));
     }
 
+    public void testDateFiltering_epochMillis() {
```
testDateFiltering_optionalFormat?

bleskes commented on core/src/test/java/org/elasticsearch/fieldstats/FieldStatsTests.java:
```diff
@@ -359,4 +361,88 @@ public void testDateFiltering() {
+
+        DateTime dateTime1 = new DateTime(2013, 12, 30, 0, 0, 0, 0, DateTimeZone.UTC);
+        DateTime dateTime2 = new DateTime(2013, 12, 31, 0, 0, 0, 0, DateTimeZone.UTC);
+        response = client().prepareFieldStats()
```
do we really need so many tests? this is just about parsing? It can probably just have unit testing for this..

martijnvg replied:
it isn't just about parsing, but also about the fact that the date format gets used. But yes the amount of tests can be less :)

bleskes commented on core/src/test/java/org/elasticsearch/fieldstats/FieldStatsTests.java:
```diff
@@ -359,4 +361,88 @@ public void testDateFiltering() {
+                .setIndexContraints(new IndexConstraint("value", MIN, GTE, String.valueOf(dateTime1.getMillis()), "epoch_millis"), new IndexConstraint("value", MAX, LTE, String.valueOf(dateTime2.getMillis()), "epoch_millis"))
+                .setLevel("indices")
+                .get();
+        assertThat(response.getIndicesMergedFieldStats().size(), equalTo(0));
```
can we also test what happens when the value is not parseable by format given?

bleskes commented:
Thx @martijnvg . Left some minor comments

martijnvg commented:
@boaz I've updated the PR based on your comments.

bleskes commented on core/src/test/resources/org/elasticsearch/action/fieldstats/fieldstats-index-constraints-request.json:
```diff
@@ -28,6 +28,16 @@
         "lt": 9
       }
+    },
+    "field1": {
```
can we have one without a format? also, for a future change, can we not call them field 1 to 5 but something with meaning like date_field?

martijnvg replied:
the other fields are already without a format, so the parsing is properly logic is properly tested.

bleskes commented on docs/reference/search/field-stats.asciidoc:
```diff
@@ -263,3 +263,25 @@ Each index constraint support the following comparisons:
 `lte`:: 	Less-than or equal to
 `lt`::  	Less-than
+
+Field stats index constraints on date fields optionally also accept a format option, that overrides the format being
```
"Field stats index constraints on date fields optionally accept a format option, used to parse the constraint's value. If missing, the format configured in the field's mapping is used.

bleskes commented:
LGTM. Left some minor comments. No need for another review.
